src/rust/engine/fs/src/lib.rs

```diff
@@ -503,6 +508,36 @@ impl PathGlobs {
       GlobExpansionConjunction::AllMatch,
     )
   }
+
+  ///
+  /// Matches these PathGlobs against the given paths.
+  ///
+  /// NB: This implementation is independent from VFS::expand, and must be kept in sync via unit
+  /// tests. The lazy filesystem traversal in VFS::expand is (currently) too expensive to use for
+  /// in-memory matching (such as via MemFS).
+  ///
```

Comment:
I don't know if I see why this needs to be kept in sync with `VFS::expand`. To me it seems perfectly natural to have a method that matches globs against literal paths, and to have that be much faster / in memory compared to expanding those globs against an arbitrary filesystem. I would at least make it clear that this method does this matching quickly and in-memory -- this comment is a little confusing to me.

If this method is *extremely* fast in comparison, one way to check the correctness of glob matching is to run this `.matches()` method on every path returned from `VFS::expand`, every time we expand a glob against the `VFS`, and raise some "internal error" message if there is a dissonance. I think this could be a good idea but that might just be me.

Also, I think it's actually `GlobMatchingImplementation::expand()` now?

Reply:
It's not clear to me what needs to be kept in sync here, or what unit tests ensure it is. Can you flesh out this comment?

Reply:
Expanded it a bit further to reference the test that it is keeping these synched.